### PR TITLE
Report/Project Client Access Tests

### PR DESCRIPTION
**Pull request by dtgreiner**
## _Merging this PR_
- use the squash-merge strategy for PRs targeting a release-X branch
- use a merge-commit or rebase strategy for PRs targeting the stable branch

## Description
[//]: # (Summarize changes and include links related issue)
[//]: # (List any new dependencies or relevant ADRs)

Adds tests for the separation of client access within report and project data. This should help verify the `Filters::Criteria::Base#viewable_project_scope` method and how it handles the separation of client access with report and project data based on the `can_view_assigned_reports` and `can_view_projects` permissions.

## Type of change
[//]: # (e.g., Bug fix, New feature, Documentation, Code clean-up, Dependency update)

 New feature - added Tests

## Checklist before requesting review
[//]: # (Remove any items that are not applicable)
- [X] I have performed a self-review of my code
- [X] I have run the code that is being changed under ideal conditions, and it doesn't fail
- [ ] If adding a new endpoint / exposing data in a new way, I have:
  - [ ] ensured the API can't leak data from other data sources
  - [ ] ensured this does not introduce N+1s
  - [ ] ensured permissions and visibility checks are performed in the right places
- [ ] Any major architectural changes are supported by an approved ADR (Architectural Decision Record) 
- [ ] I have updated the documentation (or not applicable)
- [X] I have added spec tests (or not applicable)
- [ ] I have provided testing instructions in this PR or the related issue (or not applicable)

[//]: # NOTE: system tests may fail if there is no branch on the hmis-frontend that matches the Source  or Target branch of this PR. This is expected

### CODE REVIEW

File: spec/models/filters/criteria/base_spec.rb [hunk 1]
```diff
@@ -0,0 +1,269 @@
...
+  let!(:enrollment_1) { create_enrollment_for_client(client_1, project_id: project_1.ProjectID) }
+  let!(:enrollment_2) { create_enrollment_for_client(client_2, project_id: project_2.ProjectID) }
+  let!(:enrollment_3) { create_enrollment_for_client(client_3, project_id: project_3.ProjectID) }
+  let!(:enrollment_4) { create_enrollment_for_client(client_4, project_id: confidential_project.ProjectID) }
+
+  describe '#viewable_project_scope' do
+    context 'when user has can_view_assigned_reports permission' do
+      before do
+        # Give user access to all projects for reporting
+        setup_access_control(user, role, Collection.system_collection(:data_sources))
```
Comment (eanders) [inline spec/models/filters/criteria/base_spec.rb:36]: I'm probably blind, where is role defined?

File: spec/models/filters/criteria/base_spec.rb [hunk 2]
```diff
@@ -0,0 +1,269 @@
...
+      end
+
+      it 'includes confidential projects when user can report on confidential projects' do
+        # Add confidential reporting permission
+        role.update!(can_report_on_confidential_projects: true)
+
+        scope = criteria.viewable_project_scope
+        project_ids = scope.pluck(:id)
+
+        expect(project_ids).to include(confidential_project.id)
```
Comment (eanders) [inline spec/models/filters/criteria/base_spec.rb]: Maybe this should confirm the prior 3 projects as well?  Just to confirm we're not _only_ getting back the confidential project.

File: spec/models/filters/criteria/base_spec.rb [hunk 3]
```diff
@@ -0,0 +1,269 @@
...
+
+      it 'returns no projects' do
+        scope = no_report_criteria.viewable_project_scope
+        expect(scope).to be_empty
+      end
+    end
+  end
+
+  describe 'separation between projects and reports' do
+    let(:report_role) { create(:role, can_view_assigned_reports: true, can_view_projects: true) }
```
Comment (eanders) [inline spec/models/filters/criteria/base_spec.rb]: Should this be `can_view_projects: false`?

File: spec/models/filters/criteria/base_spec.rb [hunk 4]
```diff
@@ -0,0 +1,269 @@
...
+    it 'allows running reports for organization data' do
+      report_filter = ::Filters::FilterBase.new(user_id: report_user.id)
+      report_criteria = described_class.new(input: report_filter, config: config)
+
+      scope = report_criteria.viewable_project_scope
+      project_ids = scope.pluck(:id)
+
+      # Report user can see all projects in the organization for reporting
+      expect(project_ids).to include(project_1.id, project_2.id, project_3.id)
+    end
```
Comment (eanders) [inline spec/models/filters/criteria/base_spec.rb]: I think we want to confirm the report user can't see the clients in the projects maybe?  I'm a bit confused about what it should be showing here and in the block starting on 148.  Why can't the project user see any projects?

File: spec/models/filters/criteria/base_spec.rb [hunk 5]
```diff
@@ -0,0 +1,269 @@
...
+
+      # Test client dashboard access (uses can_view_projects)
+      dashboard_scope = GrdaWarehouse::Hud::Project.viewable_by(project_user, permission: :can_view_projects)
+
+      # Report user can access all organization projects for reporting
+      expect(report_scope.pluck(:id)).to include(project_1.id, project_2.id, project_3.id)
+
+      # Project user can only access specific projects for client dashboards
+      expect(dashboard_scope.pluck(:id)).to contain_exactly(project_1.id)
+    end
```
Comment (eanders) [inline spec/models/filters/criteria/base_spec.rb]: Again, I think we want to confirm the set returned for `GrdaWarehouse::Hud::Project.viewable_by(report_user, permission: :can_view_projects)`
and
`GrdaWarehouse::Hud::Project.viewable_by(project_user, permission: :can_view_assigned_reports)`
Which, if permissions are set correctly, would both be empty.  Right now, I think `report_user` can see the projects with `can_view_projects`.

File: spec/models/filters/criteria/base_spec.rb (hunk 1 above)
Comment (dtgreiner) [reply spec/models/filters/criteria/base_spec.rb:36]: This is from the shared 'filter criteria setup' context - [shared_filter_criteria_context.rb](https://github.com/greenriver/hmis-warehouse/blob/956281bb8988c3f86ab9f0cecc51461c9c3ea0de/spec/models/filters/criteria/shared_filter_criteria_context.rb#L22)

File: spec/models/filters/criteria/base_spec.rb (hunk 3 above)
Comment (dtgreiner) [reply spec/models/filters/criteria/base_spec.rb]: yes, I must have missed this in a cppy/paste. Including it in the next commit.

File: spec/models/filters/criteria/base_spec.rb (hunk 4 above)
Comment (dtgreiner) [reply spec/models/filters/criteria/base_spec.rb]: This was attempting to be a bit of a redundancy to the previous test and use viewable_project_scope as is likely how it is used in the wild. `viewable_project_scope` is just returning the projects viewable_by with `can_view_assigned_reports` ([link](https://github.com/greenriver/hmis-warehouse/blob/956281bb8988c3f86ab9f0cecc51461c9c3ea0de/app/models/filters/criteria/base.rb#L25)) so the project user wouldn't have access here due to that permission.

I'm updating to make these tests a bit more clear and make sure they test the reverse scenarios to show the project sets are empty in those cases.

File: spec/models/filters/criteria/base_spec.rb [hunk 6]
```diff
@@ -0,0 +1,272 @@
...
+        limited_collection.set_viewables({ projects: [project_1.id, project_2.id] })
+        setup_access_control(limited_user, limited_role, limited_collection)
+      end
+
+      it 'returns only projects the user has access to' do
+        scope = limited_criteria.viewable_project_scope
+        project_ids = scope.pluck(:id)
+
+        expect(project_ids).to contain_exactly(project_1.id, project_2.id)
+        expect(project_ids).not_to include(project_3.id, confidential_project.id)
```
Comment (eanders) [inline spec/models/filters/criteria/base_spec.rb]: no real harm here, but I think `contain_exactly` covers `not_to` as well

File: spec/models/filters/criteria/base_spec.rb [hunk 7]
```diff
@@ -0,0 +1,272 @@
...
+
+    it 'report and project access are separate using permissions' do
+      # Test report access (uses can_view_assigned_reports)
+      report_scope = GrdaWarehouse::Hud::Project.viewable_by(report_user, permission: :can_view_assigned_reports)
+
+      # Test project access (uses can_view_projects)
+      project_scope = GrdaWarehouse::Hud::Project.viewable_by(project_user, permission: :can_view_projects)
+
+      # Report user can access projects for reporting
+      expect(report_scope.pluck(:id)).to include(project_1.id, project_2.id, project_3.id)
```
Comment (eanders) [inline spec/models/filters/criteria/base_spec.rb]: Should this be:
```suggestion
      expect(report_scope.pluck(:id)).to contain_exactly(project_1.id, project_2.id, project_3.id)
```

File: spec/models/filters/criteria/base_spec.rb [hunk 8]
```diff
@@ -0,0 +1,272 @@
...
+    before do
+      setup_access_control(user, role, Collection.system_collection(:data_sources))
+    end
+
+    it 'uses viewable_project_scope in FilterForUserAccess' do
+      scope = GrdaWarehouse::ServiceHistoryEnrollment.entry
+      result = filter_for_user_access.apply(scope)
+
+      # Should only include enrollments from viewable projects
+      project_ids_in_result = result.joins(:project).distinct.pluck('Project.id')
```
Comment (eanders) [inline spec/models/filters/criteria/base_spec.rb]: Not sure this pluck should work.  The case insensitivity of postgres _should_ complain...  Maybe `p_t[:id]` or "#{GrdaWarehouse::Hud::Project.quoted_table_name}.id" would be better than `pluck('Project.id')`

**Review by eanders — APPROVED** [review]
Couple of minor comments, otherwise, this looks great, thank you!
